src/System.Private.ServiceModel/src/System/ServiceModel/Channels/MessageContent.cs

```diff
@@ -0,0 +1,266 @@
+                                }
+                                catch (FormatException formatException)
+                                {
+                                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                                        new ProtocolException(SR.Format(SR.HttpContentTypeFormatException, formatException.Message, Headers.ContentType.ToString()), formatException));
+                                }
+                            }
+                        }
+
+                        if (shouldSetContentType)
+                        {
+                            Headers.ContentType.Parameters.Add(new NameValueHeaderValue("action", action));
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool SetContentType(string contentType)
+        {
+            MediaTypeHeaderValue contentTypeHeaderValue;
+            if (MediaTypeHeaderValue.TryParse(contentType, out contentTypeHeaderValue))
+            {
+                Headers.ContentType = contentTypeHeaderValue;
+                return true;
+            }
+            return false;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if(disposing && !_disposed)
+            {
+                _disposed = true;
+                if (_stream != null)
+                {
+                    var stream = _stream;
+                    _stream = null;
+                    stream.Dispose();
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        internal static HttpContent Create(HttpChannelFactory<IRequestChannel> factory, Message request, TimeoutHelper _timeoutHelper)
+        {
+            if(TransferModeHelper.IsRequestStreamed(factory.TransferMode))
+            {
+                return new StreamedMessageContent(request, factory.MessageEncoderFactory.Encoder);
+            }
+            else
+            {
+                return new BufferedMessageContent(request, factory.MessageEncoderFactory.Encoder, factory.BufferManager);
+            }
+        }
+    }
+
+    internal class StreamedMessageContent : MessageContent
+    {
+        private Task _encodingTask;
+
+        public StreamedMessageContent(Message message, MessageEncoder messageEncoder) : base(message, messageEncoder)
+        {
+        }
+
+
+        protected override Task<Stream> CreateContentReadStreamAsync()
+        {
+            _stream = new ProducerConsumerStream();
+            _encodingTask = _messageEncoder.WriteMessageAsync(_message, _stream);
+            return Task.FromResult(_stream);
+        }
+
+        protected override async Task SerializeToStreamAsync(Stream stream, TransportContext context)
+        {
+            _stream = new ProducerConsumerStream();
+            await _messageEncoder.WriteMessageAsync(_message, _stream);
+            await _stream.CopyToAsync(stream, 8192);
```

Comment:
I realized it was not needed. We can pass the original stream into WriteMessageAsync, there's no need for an intermediate stream. The passed in stream might need to be wrapped in a BufferedStream depending on it's implementation, but that will show up in perf tests later if it's an issue.